templates/default/ckan-email-notifications.sh.erb

```diff
@@ -0,0 +1,26 @@
+#!/bin/sh
+
+VIRTUAL_ENV=${VIRTUAL_ENV:-/usr/lib/ckan/default}
+CKAN_INI=${CKAN_INI:-/etc/ckan/default/production.ini}
+EMAIL_NOTIFICATION_API=/api/action/send_email_notifications
+EMAIL_NOTIFICATION_ENDPOINT=https://<%= @app_url %><%= node['datashades']['ckan_web']['endpoint'] %>$EMAIL_NOTIFICATION_API
+ACCOUNT=<%= @app_name %>
+
+if [ -f "$VIRTUAL_ENV/bin/ckan" ]; then
+  # CKAN >= 2.9
+  if ($VIRTUAL_ENV/bin/ckan_cli notify --help 2>/dev/null |grep 'send_emails' >/dev/null 2>&1); then
+    $VIRTUAL_ENV/bin/ckan_cli notify send_emails
+  else
+    # Create an API token and use that to authenticate our request
+    TOKEN_NAME=email_notification_token_$(date +'%s')
+    TOKEN=$($VIRTUAL_ENV/bin/ckan_cli user token add $ACCOUNT $TOKEN_NAME | tail -1 | tr -d '[:space:]')
```

Comment:
does this provide token id?
can't we store this on disk and do verification if token id is still against user and if not, create a new one if rotated?

Reply:
The ID is not provided at creation time, no. We instead use the token name to look it up later.

Storing on disk is possible if we want to go that route, but it would come with its own complications when coordinating amongst multiple job servers (eg if we delete and recreate one). We would need additional 'token list' calls to detect if the token is meant to have been generated already, and we would need to handle the situation where the token apparently exists but is not found on disk. (Do we revoke it, and potentially leave another server with a token that is no longer valid, which requires further logic to handle it? Do we leave it unrevoked, and potentially accumulate 'dead' tokens?)